### PR TITLE
Add/Edit or Delete a tenant feature

### PR DESCRIPTION
**Pull request by JoeChapman**
This PR assigns a Behaviour to the `/tenant-another` step to enhance the current forms with the ability to edit any one of the tenants' details or completely remove a/all tenant(s).

It's a bit big. I suggest you just check it out and use it.
Steps: 
Navigate to `/tenant-another` and start clicking away.

### CODE REVIEW

**Comment by sulthan-ahmed** [comment]
it's failing on CI

**Comment by sulthan-ahmed** [comment]
I get some strange behaviour, if I delete all entries and click the back button, I still get a blank entry on the table, see below 

screen shot 2017-08-14 at 14 56 19

File: apps/right-to-rent-check/acceptance/features/tenant-delete-last.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+Before((
+  I,
+  tenantAddAnotherPage
+) => {
+  I.visitPage(tenantAddAnotherPage, steps);
+  I.seeNumberOfElements('.tenant-details', 1);
+});
+
+Scenario('I see all fields empty on the the tenant-details page', (
```
Comment (sulthan-ahmed) [inline apps/right-to-rent-check/acceptance/features/tenant-delete-last.js]: too many `the`

File: apps/right-to-rent-check/acceptance/features/tenant-delete-last.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  I,
+  tenantAddAnotherPage,
+  tenantDetailsPage,
+  tenantAdditionalDetailsPage
+) => {
+  I.click('.tenant-details:first-of-type tr:nth-child(7) > td:nth-child(4) > a');
```
Comment (sulthan-ahmed) [inline apps/right-to-rent-check/acceptance/features/tenant-delete-last.js]: I'm assuming this is the Delete button, is there any way of making that clear e.g. maybe a comment or adding it to the associated page file?  Also I think this feature could have the kind of explanation you talked about before, 'Given I delete the last tenant on the tenant-add-another page', Scenario 'I see all fields empty on the tenant-details-page

File: apps/right-to-rent-check/acceptance/features/tenant-delete-last.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  I.seeInCurrentUrl(tenantDetailsPage.url);
+  I.seeInField('tenant-name', '');
+  I.seeInField('tenant-dob-day', '');
+  I.seeInField('tenant-dob-month', '');
+  I.seeInField('tenant-dob-year', '');
+  I.seeInField('tenant-country', '');
```
Comment (sulthan-ahmed) [inline apps/right-to-rent-check/acceptance/features/tenant-delete-last.js]: shame this doesn't accept an array :(

File: apps/right-to-rent-check/acceptance/features/tenant-delete-last.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  tenantAddAnotherPage,
+  tenantDetailsPage,
+  tenantAdditionalDetailsPage
+) => {
+  I.click('.tenant-details:first-of-type tr:nth-child(7) > td:nth-child(4) > a');
+  I.fillField('tenant-name', 'aaa');
+  I.fillField('tenant-dob-day', '1');
+  I.fillField('tenant-dob-month', '1');
+  I.fillField('tenant-dob-year', '1980');
+  I.fillField('tenant-country', 'Uganda');
```
Comment (sulthan-ahmed) [inline apps/right-to-rent-check/acceptance/features/tenant-delete-last.js]: I don't mind this here as it means I don't have to go to the page file and it's a lot quicker to do but 

is there a question of inconsistency of whether you want your data in your page file or not

File: apps/right-to-rent-check/acceptance/features/tenant-delete.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  I.fillField('#tenant-dob-month', '1');
+  I.fillField('#tenant-dob-year', '1980');
+  I.fillField('#tenant-country', 'United Kingdom');
+  I.submitForm();
+
+  I.seeInCurrentUrl('tenant-additional-details');
+  I.submitForm();
+
+  I.seeInCurrentUrl('tenant-another');
+  I.seeNumberOfElements('.tenant-details', 1);
```
Comment (sulthan-ahmed) [inline apps/right-to-rent-check/acceptance/features/tenant-delete.js]: do you need to check the values?

File: apps/right-to-rent-check/acceptance/features/tenant-delete.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+Scenario('I see the tenant I did not delete on the tenant-another page', (
+  I,
+  tenantAddAnotherPage
+) => {
+  I.click('.tenant-details:first-of-type tr:nth-child(7) > td:nth-child(4) > a');
+  I.seeInCurrentUrl(tenantAddAnotherPage.url);
+  I.seeNumberOfElements('.tenant-details', 1);
+  I.see('Name bbb', 'table > tbody > tr:nth-child(1)');
+  I.see('Date of birth 1981-02-02', 'table > tbody > tr:nth-child(2)');
+  I.see('Country of nationality India', 'table > tbody > tr:nth-child(3)');
```
Comment (sulthan-ahmed) [inline apps/right-to-rent-check/acceptance/features/tenant-delete.js]: do you also need a I.dont.see tenant for the one you've deleted

File: apps/right-to-rent-check/acceptance/features/tenant-details-backlink.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  I.visitPage(tenantAddAnotherPage, steps);
+});
+
+Scenario('I am taken to the tenant-additonal-details page', (
+  I,
+  tenantAddAnotherPage,
+  tenantAdditionalDetailsPage
+) => {
+  I.click(tenantAddAnotherPage.selectors.backLink);
+  I.seeInCurrentUrl(tenantAdditionalDetailsPage.url);
```
Comment (sulthan-ahmed) [inline apps/right-to-rent-check/acceptance/features/tenant-details-backlink.js]: What is the aim of this test with the backlink? is it because the looping has previously messed up?  I see the value of adding an entry, then going back, submitting and checking the values are still present but otherwise, I'm not sure :thinking: does that mean we need to check the backlink for all our features

File: apps/right-to-rent-check/acceptance/features/tenant-details-backlink.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+Scenario('I am taken back to the tenant-add-another page if click submit', (
+  I,
+  tenantAddAnotherPage,
+  tenantAdditionalDetailsPage
+) => {
+  I.click(tenantAddAnotherPage.selectors.backLink);
+  I.seeInCurrentUrl(tenantAdditionalDetailsPage.url);
+  I.submitForm();
+  I.seeInCurrentUrl(tenantAddAnotherPage.url);
```
Comment (sulthan-ahmed) [inline apps/right-to-rent-check/acceptance/features/tenant-details-backlink.js]: Again what is the aim of this test, is it just to test the journey because it seems like you might be adding an empty entry?

File: apps/right-to-rent-check/acceptance/features/tenant-details-change-link.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+const _ = require('lodash');
+const steps = require('../../');
+
+Feature('Tenant Add Another Page Details Change links');
+
+Before((
+  I,
+  tenantAddAnotherPage
+) => {
+  I.visitPage(tenantAddAnotherPage, steps);
```
Comment (sulthan-ahmed) [inline apps/right-to-rent-check/acceptance/features/tenant-details-change-link.js]: As before, you may need to add an actual tenant in this step as this may be just creating an empty tenant

File: apps/right-to-rent-check/acceptance/features/tenant-details.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+});
+
+Scenario('I see an error for each field if I submit without completing any of the fields', (
+  I,
+  tenantDetailsPage
+) => {
+  I.submitForm();
+  I.seeErrors(tenantDetailsPage.fields.name);
+  I.seeErrors(tenantDetailsPage.fields.country);
+  I.seeErrors(tenantDetailsPage.fields.date);
```
Comment (sulthan-ahmed) [inline apps/right-to-rent-check/acceptance/features/tenant-details.js]: This can be put in an array and also _.values

File: apps/right-to-rent-check/acceptance/features/tenant-another.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+Scenario('The correct fields are on the page', (
+  I,
+  tenantAddAnotherPage
+) => {
+  I.seeElements(_.values(tenantAddAnotherPage.fields));
+});
+
+Scenario('I see one tenant details table on the page', (
+  I,
+  tenantAddAnotherPage
```
Comment (sulthan-ahmed) [inline apps/right-to-rent-check/acceptance/features/tenant-another.js]: I don't see where you are using tenantAddAnotherPage

File: apps/right-to-rent-check/acceptance/features/tenant-another.js [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+) => {
+  I.seeNumberOfElements('.tenant-details', 1);
+});
+
+Scenario('I see a change button next to each tenant detail', (
+  I,
+  tenantAddAnotherPage
```
Comment (sulthan-ahmed) [inline apps/right-to-rent-check/acceptance/features/tenant-another.js]: I don't see where you are using tenantAddAnotherPage

**Comment by sulthan-ahmed** [comment]
We have a lot of acceptance tests here, do we need them all? 

I would definitely test where things can go wrong, like changing a tenant and deleting a tenant.  I would also add a test to delete all entries, going back and then submitting the form to ensure there isn't an empty tenant entry

File: apps/right-to-rent-check/acceptance/features/tenant-delete.js (hunk 6 above)
Comment (JoeChapman) [inline apps/right-to-rent-check/acceptance/features/tenant-delete.js]: not really because ` I.seeNumberOfElements('.tenant-details', 1);` asserts there is only one

File: apps/right-to-rent-check/acceptance/features/tenant-delete.js (hunk 5 above)
Comment (JoeChapman) [inline apps/right-to-rent-check/acceptance/features/tenant-delete.js]: 👀 ⬇️

Comment (sulthan-ahmed) [inline apps/right-to-rent-check/acceptance/features/tenant-delete.js]: 👍

File: apps/right-to-rent-check/acceptance/features/tenant-delete.js (hunk 6 above)
Comment (sulthan-ahmed) [inline apps/right-to-rent-check/acceptance/features/tenant-delete.js]: 👍

**Comment by JoeChapman** [comment]
> We have a lot of acceptance tests here, do we need them all?

I'm not sure I understand the question.

File: apps/right-to-rent-check/acceptance/features/tenant-delete-last.js (hunk 4 above)
Comment (JoeChapman) [inline apps/right-to-rent-check/acceptance/features/tenant-delete-last.js]: Perhaps

**Comment by sulthan-ahmed** [comment]
I've messaged you about it on slack

**Comment by sulthan-ahmed** [comment]
Table alignment looks a bit off.  You can tell me this is out of scope of this PR as it's making it BIIIIG

screen shot 2017-08-16 at 15 25 03

**Comment by JoeChapman** [comment]
@sulthan-ahmed as discussed, the look and feel is out of scope and I will list the reasons why.
1) the design/look and feel is not mentioned once in the story ticket.
2) there is a screen shot attached to the ticket, but we have all discussed on many occasions why we don't use attached designs prototype as a canonical reference to the app's functionality or it's design because...
3) those things change based on user research and testing. 
And 4) they have indeed changed during the development of this code.

**Comment by sulthan-ahmed** [comment]
that's fine, do we have a separate ticket for the design?

**Comment by sulthan-ahmed** [comment]
On a separate note, I agree most of hof forms do not need a screenshot.  However, how do you provide a design (or a look and feel) of something that is not common e.g. this table without a screenshot or some kind of mockup?
How else would you test it?  Anyway, that is out of scope of this PR and this discussion

File: apps/right-to-rent-check/behaviours/tenants.js [hunk 13]
```diff
@@ -12,31 +13,106 @@ module.exports = fields => {
...
+        let tenants = req.sessionModel.get('tenants') || [];
+
+        const savedByUuid = _.find(tenants, {
+          'tenant-uuid': values['tenant-uuid']
+        });
+
+        // Have we just finished editting a tenant record
+        // or is our savedByUuid different from the current tenant
+        const editted = _.find(tenants, {edit: true}) ||
+          (!_.isEqual(_.omit(savedByUuid, 'edit', 'tenant-uuid'), tenant) ? savedByUuid : false);
```
Comment (sulthan-ahmed) [inline apps/right-to-rent-check/behaviours/tenants.js]: This is cool and works, it's just very hard to understand.  It would be good if this could be broken up a bit more

File: package.json [hunk 14]
```diff
@@ -3,10 +3,13 @@
   "scripts": {
     "start": "node app.js",
     "dev": "hof-build watch",
-    "test": "NODE_ENV=test npm run test:unit && npm run test:lint",
+    "dev:ci": "MOCK_POSTCODE=true NODE_ENV=ci DISABLE_CSP=true npm run dev",
```
Comment (sulthan-ahmed) [inline package.json]: good idea 👍

File: apps/right-to-rent-check/acceptance/features/tenant-another.js [hunk 15]
```diff
@@ -0,0 +1,270 @@
+'use strict';
+
+const _ = require('lodash');
+const steps = require('../../');
+const deleteBtn = '.tenant-details:first-of-type tr:nth-child(7) > td:nth-child(4) > a';
+
+Feature('Given I have added one tenant and /tenant-another loads');
```
Comment (JoeChapman) [inline apps/right-to-rent-check/acceptance/features/tenant-another.js]: Yeah I only wrote `and` because the song goes `Given`, `When` `Then`, not `Given`, `Then`, `When`, `Then`

File: apps/right-to-rent-check/acceptance/features/tenant-another.js [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+  I.fillField('tenant-dob-day', '2');
+  I.fillField('tenant-dob-month', '2');
+  I.fillField('tenant-dob-year', '1965');
+  I.fillField('tenant-country', 'Uganda');
+
+  I.submitForm();
+  I.seeInCurrentUrl(tenantAdditionalDetailsPage.url);
+});
+
+Scenario('When I click the change \'Name\' button Then I am redirected to /tenant-details', (
```
Comment (JoeChapman) [inline apps/right-to-rent-check/acceptance/features/tenant-another.js]: I'm trying to keep words to a minimum

File: apps/right-to-rent-check/behaviours/tenants.js [hunk 17]
```diff
@@ -12,31 +13,106 @@ module.exports = fields => {
...
-        const tenants = req.sessionModel.get('tenants') || [];
-        // Find a tenant matching the current tenant
-        let tenant = _.find(tenants, picked);
-
-        // Add a new tenant
-        if (!tenant) {
-          tenant = _.cloneDeep(picked);
-          if (!_.isEmpty(tenant)) {
-            tenants.push(tenant);
+        let tenant = _.pick(values, fields);
```
Comment (JoeChapman) [inline apps/right-to-rent-check/behaviours/tenants.js]: I don't think we should annotate where there is already documentation online: https://lodash.com/docs/4.17.4#pick

File: apps/right-to-rent-check/acceptance/features/tenant-additional-details.js [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+  I.seeElements(_.values(tenantAdditionalDetailsPage.hidden));
+});
+
+Scenario('Hidden fields become visible when the fields are checked', (
+  I,
+  tenantAdditionalDetailsPage
+) => {
+  tenantAdditionalDetailsPage.checkFields();
+  I.seeElements(_.values(tenantAdditionalDetailsPage.hidden));
+});
```
Comment (JoeChapman) [inline apps/right-to-rent-check/acceptance/features/tenant-additional-details.js]: no

File: apps/right-to-rent-check/acceptance/features/tenant-details.js [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+  I.seeElements(_.values(tenantDetailsPage.fields));
+});
+
+Scenario('When I submit an empty form then I see an error for each field', (
+  I
+) => {
+  I.submitForm();
+  I.seeErrors('#tenant-name');
+  I.seeErrors('#tenant-country');
+  I.seeErrors('#tenant-dob-group');
```
Comment (JoeChapman) [inline apps/right-to-rent-check/acceptance/features/tenant-details.js]:
```
I.seeErrors(['#tenant-name', '#tenant-dob', '#tenant-country']));
```

File: apps/right-to-rent-check/acceptance/features/tenant-delete.js [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+const steps = require('../../');
+
+Feature('Given I have added two tenants');
+
+Before((
+  I,
+  tenantDetailsPage
+) => {
+  I.visitPage(tenantDetailsPage, steps);
+
```
Comment (JoeChapman) [inline apps/right-to-rent-check/acceptance/features/tenant-delete.js]: what paragraph?

Comment (JoeChapman) [inline apps/right-to-rent-check/acceptance/features/tenant-delete.js]: do you mean new line?

Comment (JoeChapman) [inline apps/right-to-rent-check/acceptance/features/tenant-delete.js]: it's intentional, I think it's easier to understand the evens

**Comment by JoeChapman** [comment]
@sulthan-ahmed @easternbloc https://media.giphy.com/media/104ueR8J1OPM2s/giphy_s.gif

File: apps/right-to-rent-check/acceptance/pages/tenant-another.js [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+  fields: {
+    group: '#tenant-add-another-group',
+    yes: '#tenant-add-another-yes',
+    no: '#tenant-add-another-no'
+  },
+
+  selectors: {
+    backLink: '#step a',
+    change: {
+      name: '.tenant-details tr:nth-child(1) > td:nth-child(3) > a',
```
Comment (easternbloc) [inline apps/right-to-rent-check/acceptance/pages/tenant-another.js]: this i'd love to have something nicer than all these...

File: apps/right-to-rent-check/views/tenant-another.html [hunk 22]
```diff
@@ -5,37 +5,47 @@
...
+    {{#hasTenants}}
+      <h2>{{#t}}pages.tenant-another.tables.title{{/t}}</h2>
+    {{/hasTenants}}
     {{#tenants}}
       <table class="tenant-details">
         <tbody>
           <tr>
             <th>{{#t}}pages.tenant-another.tables.headers.name{{/t}}</th>
             <td>{{tenant-name}}</td>
+            <td><a href="/tenant-another/edit/{{tenant-uuid}}?field=tenant-name">{{#t}}buttons.change{{/t}}</td>
```
Comment (easternbloc) [inline apps/right-to-rent-check/views/tenant-another.html]: change-tenant-name in the class it'll make your tests less fragile and more readable.

rinse and repeat for each one

File: apps/right-to-rent-check/acceptance/features/tenant-additional-details.js [hunk 23]
```diff
@@ -0,0 +1,70 @@
+'use strict';
+
+const _ = require('lodash');
+const steps = require('../../');
+
+Feature('Given I am at /tenant-additional-details');
```
Comment (JoeChapman) [inline apps/right-to-rent-check/acceptance/features/tenant-additional-details.js]: the urls' are canonical. However, things we call pages changes and is different from person to person. In my view urls are more reliable. It is not bad

File: apps/right-to-rent-check/acceptance/features/tenant-delete.js (hunk 20 above)
Comment (sulthan-ahmed) [inline apps/right-to-rent-check/acceptance/features/tenant-delete.js]: ok cool

File: apps/right-to-rent-check/acceptance/features/tenant-details.js (hunk 19 above)
Comment (sulthan-ahmed) [inline apps/right-to-rent-check/acceptance/features/tenant-details.js]: yeah

File: apps/right-to-rent-check/behaviours/tenants.js (hunk 17 above)
Comment (sulthan-ahmed) [inline apps/right-to-rent-check/behaviours/tenants.js]: no problem at all

**Comment by sulthan-ahmed** [comment]
something is failing

**Comment by JoeChapman** [comment]
@sulthan-ahmed what does that mean?

**Comment by sulthan-ahmed** [comment]
The build was failing but it seems fine now

**Comment by JoeChapman** [comment]
@sulthan-ahmed as always your invaluable feedback has been most appreciated

**Comment by JoeChapman** [comment]
@easternbloc and @sulthan-ahmed thank you both for your hard work reviewing this PR. I have done everything reasonable that was asked of me. Please can you merge this PR
